### PR TITLE
Test: Support module mocking with conditional subpath imports in `package.json` 

### PR DESCRIPTION
**Pull request by kasperpeulen**
## What I did

* This adds storybook, stories and the test condition to the webpack and vite builder. Allowing user to do conditional import maps in package.json to be used for module mocking use cases.
* https://github.com/storybookjs/storybook/pull/26750
* https://github.com/storybookjs/storybook/pull/26770
* https://github.com/storybookjs/storybook/pull/26815
* https://github.com/storybookjs/storybook/pull/26864
* https://github.com/storybookjs/storybook/pull/26853
* https://github.com/storybookjs/storybook/pull/26789

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [x] unit tests
- [x] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-26688-sha-553a642b`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-26688-sha-553a642b sandbox` or in an existing project with `npx storybook@0.0.0-pr-26688-sha-553a642b upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-26688-sha-553a642b`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-26688-sha-553a642b) |
| **Triggered by** | @yannbf |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`kasper/module-mocking`](https://github.com/storybookjs/storybook/tree/kasper/module-mocking) |
| **Commit** | [`553a642b`](https://github.com/storybookjs/storybook/commit/553a642b8e9657152c7b66fe4a7467719db09521) |
| **Datetime** | Mon Apr 15 07:21:36 UTC 2024 (`1713165696`) |
| **Workflow run** | [8685051027](https://github.com/storybookjs/storybook/actions/runs/8685051027) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=26688`_
</details>

### CODE REVIEW

**Review by JReinhold — APPROVED** [review]
I didn't go through all the code _again_, just the important tests and stories.

Looks great 💪

File: code/lib/test/template/stories/before-each.stories.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { global as globalThis } from '@storybook/global';
```
Comment (yannbf) [inline code/lib/test/template/stories/before-each.stories.ts:1]: Shouldn't we just stop using @storybook/global entirely and just use globalThis directly?
```suggestion
```

File: code/lib/test/template/stories/module-mocking.stories.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  args: {
+    onClick: fn(),
+    label: 'Mock story',
+  },
+  parameters: {
+    chromatic: {
+      disable: true,
+    },
+  },
+  loaders: () => {
```
Comment (yannbf) [inline code/lib/test/template/stories/module-mocking.stories.ts:19]: is loaders used on purpose, or did you mean to use beforeEach?

File: code/builders/builder-vite/src/vite-config.ts [hunk 3]
```diff
@@ -65,6 +65,7 @@ export async function commonConfig(
     base: './',
     plugins: await pluginConfig(options),
     resolve: {
+      conditions: ['storybook', 'stories', 'test'],
```
Comment (yannbf) [inline code/builders/builder-vite/src/vite-config.ts:68]: Is this going to be merged correctly? Or does it override the user's base config? If so, we need to change this.

File: code/addons/actions/src/containers/ActionLogger/index.tsx [hunk 4]
```diff
@@ -63,12 +63,12 @@ export default class ActionLogger extends Component<ActionLoggerProps, ActionLog
...
     this.setState((prevState: ActionLoggerState) => {
       const actions = [...prevState.actions];
-      const previous = actions.length && actions[0];
+      const previous = actions.length && actions[actions.length - 1];
       if (previous && safeDeepEqual(previous.data, action.data)) {
         previous.count++;
       } else {
         action.count = 1;
-        actions.unshift(action);
+        actions.push(action);
```
Comment (yannbf) [inline code/addons/actions/src/containers/ActionLogger/index.tsx:71]: If we haven't yet, we need to document this change

**Review by yannbf — APPROVED** [review]
LGTM apart from the Vite conditions change I made a comment of

File: code/lib/preview-api/template/stories/rendering.stories.ts [hunk 5]
```diff
@@ -66,3 +66,31 @@ export const ChangeArgs = {
...
+
+let loadedLabel = 'Initial';
+
+/**
+ * This story demonstrates what happens when rendering (loaders) have side effects, and can possibly interleave with each other
+ * Triggering multiple force remounts quickly should only result in a single remount in the end
+ * and the label should be 'Loaded. Click Me' at the end. If loaders are interleaving it would result in a label of 'Error: Interleaved loaders. Click Me'
+ * Similarly, changing args rapidly should only cause one rerender at a time, producing the same result.
+ */
+export const SlowLoader = {
```
Comment (kasperpeulen) [inline code/lib/preview-api/template/stories/rendering.stories.ts:78]: Shall I disable this one in chromatic, as it is e2e tested and hard to judge if it regressed in chromatic?

File: code/builders/builder-vite/src/vite-config.ts (hunk 3 above)
Comment (JReinhold) [reply code/builders/builder-vite/src/vite-config.ts:68]: We merge this with the user config using Vite's `mergeConfig`, so I think we're good.

https://vitejs.dev/guide/api-javascript#mergeconfig

https://github.com/storybookjs/storybook/blob/kasper/module-mocking/code/builders/builder-vite/src/vite-config.ts#L83

File: code/lib/test/template/stories/module-mocking.stories.ts (hunk 2 above)
Comment (kasperpeulen) [reply code/lib/test/template/stories/module-mocking.stories.ts:19]: beforeEach was not available when I made this story! will change :)

File: code/lib/preview-api/template/stories/rendering.stories.ts (hunk 5 above)
Comment (JReinhold) [reply code/lib/preview-api/template/stories/rendering.stories.ts:78]: Sounds good to me.

File: code/addons/actions/src/containers/ActionLogger/index.tsx (hunk 4 above)
Comment (kasperpeulen) [reply code/addons/actions/src/containers/ActionLogger/index.tsx:71]: You mean in the migration note?

File: code/builders/builder-vite/src/vite-config.ts (hunk 3 above)
Comment (kasperpeulen) [reply code/builders/builder-vite/src/vite-config.ts:68]: Yes, I also manually tested that it doesn't remove base conditions like development and browser. Also the docs say:
> Vite has a list of "allowed conditions" and will match the first condition that is in the allowed list. The default allowed conditions are: import, module, browser, default, and production/development based on current mode. The resolve.conditions config option allows specifying *additional* allowed conditions.
https://vitejs.dev/config/shared-options#resolve-conditions

File: code/addons/actions/src/containers/ActionLogger/index.tsx (hunk 4 above)
Comment (kasperpeulen) [reply code/addons/actions/src/containers/ActionLogger/index.tsx:71]: @kylegach Do we have existing docs that might need to be updated?

File: code/builders/builder-vite/src/vite-config.ts (hunk 3 above)
Comment (yannbf) [reply code/builders/builder-vite/src/vite-config.ts:68]: Sounds good!

**Pull request merged**